### PR TITLE
Integrate PMD #751

### LINKED ISSUE #751: Try findbugs on HT
We can give this a try http://findbugs.sourceforge.net/

### PR DESCRIPTION
**Pull request by dariusf**
Fixes #751.

This integrates PMD, a static analysis tool for source code (FindBugs works on bytecode, so they are better at picking up different kinds of problems).

The more widely-applicable rulesets are enabled (basic, empty, unnecessary, braces, strings, design, unusedcode; full set [here](https://pmd.github.io/pmd-5.4.1/pmd-java/rules/index.html)). About 700 violations were fixed. Future work would involve enabling more and fixing the violations that result.

Implementation: there are two custom rulesets (for tests and source code) which pull in the used rulesets and define exceptions. In general exceptions for tests and source code may differ: for example, empty catch blocks are acceptable in tests, but are generally a bad sign in source code.

Usage: there are different ways to suppress warnings for false positives.

- `// NOPMD` line comments. These disable PMD for specific lines.
- `@SuppressWarnings("PMD")` annotations. These disable rules at the class or method level.
- Exceptions in rulesets. This is used to turn off entire rules.

Suppressed warnings should generally come with an explanatory comment if the reasoning isn't obvious (to reviewers).

### CODE REVIEW

**Comment by dariusf** [comment]
@ndt93 could you help to review? If the reason for any of the changes is unclear I can add comments or explain.

Also, I'm not sure if that last commit is the right way to fix that test.

Coverage drop is due to the new assert [here](https://github.com/HubTurbo/HubTurbo/blob/751-integrate-pmd/src/main/java/backend/stub/DummyRepo.java#L86), which I don't think we're going to cover.

**Comment by ndt93** [comment]
I've commented on the last commit. All other refactoring improve the code's details overall so they shouldn't be a problem.

**Comment by ndt93** [comment]
All right, no more comments from me. Should be good to merged if all unstable tests have passed too.

File: src/main/java/filter/lexer/Lexer.java [hunk 1]
```diff
@@ -27,6 +27,7 @@
             new Rule("\\)", TokenType.RBRACKET),
             new Rule("\\\"", TokenType.QUOTE),
             new Rule(",", TokenType.COMMA),
+        new Rule("%", TokenType.PERCENT),
```
Comment (HansNewbie) [inline src/main/java/filter/lexer/Lexer.java]: Indentation :stuck_out_tongue:

**Comment by HansNewbie** [comment]
Any need to update documentation on PMD suppression?

**Comment by dariusf** [comment]
There was never much documentation before, but I've added the details in the first post to a new file.

**Comment by damithc** [comment]
I think some tools such as PMD are missing from the dev doc section called 'Tool stack'

**Comment by dariusf** [comment]
They are documented [here](https://github.com/HubTurbo/HubTurbo/blob/master/docs/staticAnalysis.md) now. Maybe we could put a link to that page in the Tool Stack section?

**Comment by damithc** [comment]
In fact may be we should rejig the dev docs a little. Let me open another issue for it.

**Pull request merged**